### ISSUE TITLE
Crucible backend has trouble with pointers on joins

### ISSUE THREAD

**Issue by TomMD** [GaloisInc/saw-script#213]
Consider the below code which takes a `foo **p`, allocates `*p=alloc()` then allocates again `if(b) *p = alloc()`.  Something about this pattern triggers an issue.

```
#include <stdlib.h>
#include <stdint.h>

struct foo_st {
      int x;
};

typedef struct foo_st foo;

void inner(foo **val)
{
      *val = malloc(sizeof(foo));
      (*val)->x = 1;
}

void top(size_t arrlen, foo **val)
{
      *val = malloc(sizeof(foo));
      (*val)->x = 1;
}

int outter(size_t arrlen, foo **val)
{
      foo *tmp;
      int ret=0;
      inner(&tmp);

      if(arrlen > 0) {
          top(arrlen, &tmp);
      }
      *val = tmp;
      tmp = NULL;

      return ret;
}
```

The SAW code includes _no_ override for `inner`, an override for `top` which just states the allocation and int assignment, and tries to prove that allocation and assignment post condition for `outter` when the input is sufficient for the condition to be true.

```
load_crucible_llvm_module "foo.bc";

top_ov <- crucible_llvm_verify "top"
    []
    false
    do {
        v   <- crucible_alloc (llvm_type "%struct.foo_st*");
        arrlen <- crucible_fresh_var "arrlen" (llvm_int 64);
        crucible_execute_func [crucible_term arrlen,v];

        val <- crucible_alloc (llvm_struct "struct.foo_st");
        crucible_points_to v val;
        crucible_points_to (crucible_elem val 0) (crucible_term {{ 1 : [32] }});
    }
    z3;

crucible_llvm_verify "outter"
    [top_ov]
    false
    do {
        v <- crucible_alloc (llvm_type "%struct.foo_st*");
        arrlen <- crucible_fresh_var "arrlen" (llvm_int 64);
        crucible_precond {{ arrlen == (1 : [64]) }};
        crucible_execute_func [crucible_term arrlen,v];
        val <- crucible_alloc (llvm_struct "struct.foo_st");
        crucible_points_to v val;
        crucible_points_to (crucible_elem val 0) (crucible_term {{ 1 : [32] }});
        crucible_return (crucible_term {{ 0 : [32] }});
    }
    z3;
```

The result is:

```
% clang -g -c -emit-llvm foo.c  && saw foo.saw
Loading module Cryptol
Loading file "/home/tommd/badsaw/foo.saw"
Proof succeeded! @top
Registering override for `top`
Executing override for `top`
executePostAllocation: AllocIndex 1 { i32 }
saw: Attempted to compare pointers from different allocations
in outter at :36:5
```

Some fun notes:

* Inlining the function `inner` in C makes this problem disappear.
* Removing the `if` statement makes this problem disappear.
* The problem persists with the `if` condition of `arrlen > 0 || arrlen == 0` (always true, unsigned value).

@brianhuffman The pointer compare issue as discussed.

**Comment by TomMD** [comment]
Even simpler:

```
#include <stdlib.h>

struct foo_st {
    int x;
};
typedef struct foo_st foo;

void inner(foo **val)
{
    *val = malloc(sizeof(foo));
    (*val)->x = 1;
}

int outter(size_t arrlen, foo **val)
{
    foo *tmp;
    int ret=0;

    inner(&tmp);
    free(tmp);
    // tmp = NULL;

    if(arrlen > 0 || arrlen == 0) {
        inner(&tmp);
    }
    *val = tmp;
    return ret;
}
```

And:

```
load_crucible_llvm_module "foo.bc";

crucible_llvm_verify "outter"
    []
    false
    do {
        v <- crucible_alloc (llvm_type "%struct.foo_st*");
        arrlen <- crucible_fresh_var "arrlen" (llvm_int 64);
        crucible_precond {{ arrlen == (1 : [64]) }};
        crucible_execute_func [crucible_term arrlen,v];
        val <- crucible_alloc (llvm_struct "struct.foo_st");
        crucible_points_to v val;
        crucible_points_to (crucible_elem val 0) (crucible_term {{ 1 : [32] }});
        crucible_return (crucible_term {{ 0 : [32] }});
    }
    z3;
```

Notice this version does not have the memory leak and uncommenting the `NULL` yields a new error:

```
Verifier.SAW.Simulator.Prims.iteOp: malformed arguments: 16 0)
```

**Comment by TomMD** [comment]
Oh, I should add that using SAT branches (`false` ~> `true` in the .saw file) successfully avoids this bug. As a result I'm continuing with life but it's just non-ideal to us a performance-intended option to avoid bugs.

**Comment by brianhuffman** [comment]
Fixed in GaloisInc/crucible@196fbdcb.
